File: 프로그래머스/3/43164. 여행경로/여행경로.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

vector<vector<string>> ticket; 
vector<string> answer;
bool v[100005], find_ans;
int sz;

void dfs(string start, int cnt){
    answer.push_back(start);
    if(cnt == sz) {
        find_ans = 1;
        return;
    }
    for(int i = 0; i < sz; i++){
        if(v[i] || ticket[i][0] != start) continue;
        v[i] = 1;
        dfs(ticket[i][1], cnt+1);
        if(find_ans) return;
        answer.pop_back();
        v[i] = 0;
    }
}
vector<string> solution(vector<vector<string>> tickets) {
    sort(tickets.begin(), tickets.end());
    ticket = tickets;
    sz = ticket.size();
    
    dfs("ICN", 0);
    
    return answer;
}

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    List<String> list = new ArrayList<>();
    boolean[] v;
    boolean found = false;
    int n = 0;
    
    public String[] solution(String[][] tickets) {
        n = tickets.length;
        v = new boolean[n];
        
        Arrays.sort(tickets, (o1, o2) -> {
            if(o1[0].toString().equals(o2[0]))
                return o1[1].toString().compareTo(o2[1].toString());
            
            return o1[0].toString().compareTo(o2[0].toString());
        });
            
        dfs(0, "ICN", tickets);
        
        return list.toArray(new String[list.size()]);
    }
    
    public void dfs(int cnt, String cur, String[][] tickets){
        list.add(cur);
        
        if(cnt == n) {
            found = true;
            return;
        }
        
        for(int i = 0; i < n; i++) {
            if(v[i] || !tickets[i][0].equals(cur)) continue;
            v[i] = true;
            dfs(cnt + 1, tickets[i][1], tickets); 
            if(found) return;
            v[i] = false;
            list.remove(list.size() - 1);
        }
    }
}